src/govuk/components/hide-this-page/hide-this-page.mjs

```diff
@@ -0,0 +1,65 @@
+import { nodeListForEach } from '../../common.mjs'
+
+function HideThisPage ($module) {
+  this.$module = $module
+  this.firstButton = this.$module[0].querySelector('.govuk-js-hide-this-page-button')
+  this.updateSpan = null
+  this.escCounter = 0
+  this.escTimerActive = false
+}
+
+HideThisPage.prototype.initUpdateSpan = function () {
+  this.updateSpan = document.createElement('span')
+  this.updateSpan.setAttribute('aria-live', 'polite')
+  this.updateSpan.setAttribute('class', 'govuk-visually-hidden')
+
+  this.$module[0].appendChild(this.updateSpan)
+}
+
+HideThisPage.prototype.initButtonClickHandler = function () {
+  // We put the loop here instead of in all.js because we want to have both
+  // listeners on the individual buttons for clicks and a single listener for
+  // the keyboard shortcut
+  nodeListForEach(this.$module, function ($button) {
+    $button.querySelector('.govuk-js-hide-this-page-button').addEventListener('click', this.exitPage.bind(this))
+  }.bind(this))
+}
+
+HideThisPage.prototype.exitPage = function (e) {
+  // we don't do anything with this just yet
+}
+
+HideThisPage.prototype.handleEscKeypress = function (e) {
+  if (e.key === 'Alt' || e.keyCode === 27 || e.which === 27) {
+    this.escCounter += 1
+
+    if (this.escCounter >= 3) {
+      this.escCounter = 0
+      this.firstButton.click()
+    } else {
+      this.updateSpan.innerText = 'Exit this Page key press ' + this.escCounter + ' of 3'
+    }
+
+    this.setEscTimer()
+  }
+}
+
+HideThisPage.prototype.setEscTimer = function () {
+  if (!this.escTimerActive) {
+    this.escTimerActive = true
+
+    setTimeout(function () {
+      this.escCounter = 0
+      this.escTimerActive = false
+      this.updateSpan.innerText = ''
+    }.bind(this), 2000)
+  }
+}
+
+HideThisPage.prototype.init = function () {
+  this.initUpdateSpan()
+  this.initButtonClickHandler()
+  window.addEventListener('keydown', this.handleEscKeypress.bind(this))
```

Comment:
Nice work with all this 😊 

Just a tiny tiny tiny implementation detail

Default phase for listeners is to grab events "bubbling up" (like we have here)
https://www.w3.org/TR/DOM-Level-3-Events/#event-flow

1. Capturing phase – the event goes down to the element.
2. Target phase – the event reached the target element.
3. Bubbling phase – the event bubbles up from the element.

But we're listening at the top on `window`

The `(e)` event might have already "bubbled up" through hundreds of DOM elements—potentially being cancelled/stopped by other listeners via [`e.preventDefault()`](https://developer.mozilla.org/en-US/docs/Web/API/Event/preventDefault) or [`e.stopPropagation()`](https://developer.mozilla.org/en-US/docs/Web/API/Event/stopPropagation) along the way

I'd recommend changing to the `useCapture` capturing phase (still on `window`) to ensure we grab the event on its way down, before other listeners get chance to cancel it once it bubbles back up

Reply:
The [GOV.UK Publishing Components **Modal dialogue**](https://github.com/alphagov/govuk_publishing_components/blob/main/app/assets/javascripts/govuk_publishing_components/components/modal-dialogue.js#L57) (good example) also listens for the escape key, but you can see they add the `useCapture` flag

They're good citizens and don't stop the event, but other listeners might do

Reply:
Aha I missed this, why did I set this to listen on window 🤔